w3sper.js/src/transaction.js

```diff
@@ -10,53 +10,149 @@ export const TRANSFER =
 import { AddressSyncer } from "./network/syncer/address.js";
 import { Gas } from "./network/gas.js";
 import * as ProtocolDriver from "./protocol-driver/mod.js";
-import { ProfileGenerator } from "./profile.js";
+import { ProfileGenerator, Profile } from "./profile.js";
 
-export class TransactionBuilder {
-  #bookkeeper;
+const _attributes = Symbol("builder::attributes");
 
-  #from;
-  #to;
-  #amount;
-  #obfuscated = false;
-  #gas;
+class BasicTransfer {
+  [_attributes];
 
-  constructor(bookkeeper) {
-    this.#bookkeeper = bookkeeper;
+  constructor(from) {
+    this[_attributes] = Object.create(null);
 
-    this.#gas = new Gas();
+    const value = from instanceof Profile ? { profile: from } : from;
+
+    Object.defineProperty(this, "bookentry", {
+      value,
+    });
+
+    this[_attributes].gas = new Gas();
+  }
+
+  get attributes() {
+    return { ...this[_attributes] };
   }
 
-  from(identifier) {
-    this.#from = identifier;
+  amount(value) {
+    this[_attributes].amount = value;
     return this;
   }
 
-  to(identifier) {
-    this.#to = identifier;
+  gas(value) {
+    this[_attributes].gas = new Gas(value);
     return this;
   }
+}
 
-  amount(value) {
-    this.#amount = value;
+export class Transfer extends BasicTransfer {
+  constructor(from) {
+    super(from);
+  }
+
+  to(identifier) {
+    let builder;
+    switch (ProfileGenerator.typeOf(String(identifier))) {
+      case "account":
+        builder = new AccountTransfer(this.bookentry);
+        break;
+      case "address":
+        builder = new AddressTransfer(this.bookentry);
+        break;
+      default:
+        throw new TypeError("Invalid identifier");
+    }
+    this[_attributes].to = identifier;
+    builder[_attributes] = this.attributes;
+
+    return builder;
+  }
+}
+
+class AccountTransfer extends Transfer {
+  constructor(from) {
+    super(from);
+  }
+
+  chain(value) {
+    this[_attributes].chain = value;
     return this;
   }
 
-  obfuscated() {
-    this.#obfuscated = true;
+  nonce(value) {
+    this[_attributes].nonce = value;
     return this;
   }
 
-  gas(value) {
-    this.#gas = new Gas(value);
+  async build(network) {
+    const sender = this.bookentry.profile;
+    const { attributes } = this;
+    const { to: receiver, amount: transfer_value, gas } = attributes;
+
+    // Obtain the chain id
+    let chainId;
+    if (!isNaN(+attributes.chain)) {
```

Comment:
Maybe you can get rid of a code branch and the double conversion by just assigning the value to `chainId` and then checking for `NaN`?

Reply:
The problem here is that `attributes.chain` needs to be the one with highest prio, so it shouldn't be the default one. Otherwise I will have a code like this:

```js
let chainId = +attributes.chain;

if (isNaN(chainId) && network) {
  ({ chainId } = await network.node.info);
} else if (isNaN(chainId)) {
  throw new Error("Chain ID is required.");
}
```

Or nested `if`, ternary operator, or similar, and I'm not sure it's better. The conversion is cheap, I opted to have the branches following the prio order,  as same level siblings `if`s.

Reply:
Very minor thing anyway: as you say the conversion is very cheap.
I tend to keep branches at a minimum as a matter of preference, sometimes even at the expense of having a slightly more complicated condition.

Reply:
Yes, but in this case you do not reduce the number of branches, you just avoid to do one conversion.
Another way to avoid conversion would be adding an additional variable, but again, since it's cheap I don't see the point.